src/generate.ts

```diff
@@ -0,0 +1,192 @@
+import type {
+	AbstractSqlField,
+	AbstractSqlModel,
+	AbstractSqlTable,
+	InNode,
+	Relationship,
+	RelationshipInternalNode,
+	RelationshipLeafNode,
+} from '@balena/abstract-sql-compiler';
+import {
+	odataNameToSqlName,
+	sqlNameToODataName,
+} from '@balena/odata-to-abstract-sql';
+import { replaceResultTransformer, TemplateTag } from 'common-tags';
+
+type RequiredModelSubset = Pick<
+	AbstractSqlModel,
+	'tables' | 'relationships' | 'synonyms'
+>;
+
+const typeHelpers = `import type { Types } from '@balena/abstract-sql-to-typescript';\n`;
+
+const trimNL = new TemplateTag(
+	replaceResultTransformer(/^[\r\n]*|[\r\n]*$/g, ''),
+);
+
+const modelNameToCamelCaseName = (s: string): string =>
+	s
+		.split(/[ -]/)
+		.map((p) => p[0].toLocaleUpperCase() + p.slice(1))
+		.join('');
+
+const getReferencedInterface = (modelName: string, mode: Mode) =>
+	`${modelNameToCamelCaseName(modelName)}['${mode}']`;
+
+const sqlTypeToTypescriptType = (
+	m: RequiredModelSubset,
+	f: AbstractSqlField,
+	mode: Mode,
+): string => {
+	if (!['ForeignKey', 'ConceptType'].includes(f.dataType) && f.checks) {
+		const inChecks = f.checks.find(
+			(checkTuple): checkTuple is InNode => checkTuple[0] === 'In',
+		);
+		if (inChecks) {
+			const [, , ...allowedValues] = inChecks;
+			return allowedValues
+				.map(([type, value]) => (type === 'Text' ? `'${value}'` : value))
+				.join(' | ');
+		}
+	}
+
+	switch (f.dataType) {
+		case 'ConceptType':
+		case 'ForeignKey': {
+			const referencedInterface = getReferencedInterface(
+				m.tables[f.references!.resourceName].name,
```

Comment:
Should we instead check if `f.references` is null and throw a more understandable error here?

Reply:
I think in practice it can never be null, it's just that typing that is pretty finicky and a bit of a pain to deal with. I'll add a safety check though